# Exploratory Data Analysis

In [1]:
import pandas as pd

In [2]:
movies = pd.read_csv("../data/movies.csv")
ratings = pd.read_csv("../data/ratings.csv")
tags = pd.read_csv("../data/tags.csv")

display(movies.shape, ratings.shape, tags.shape)
display(movies.head())
display(ratings.head())
display(tags.head())

(86537, 3)

(33832162, 4)

(2328315, 4)

,movieId,title,genres
0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
1,2,Jumanji (1995),Adventure|Children|Fantasy
2,3,Grumpier Old Men (1995),Comedy|Romance
3,4,Waiting to Exhale (1995),Comedy|Drama|Romance
4,5,Father of the Bride Part II (1995),Comedy


,userId,movieId,rating,timestamp
0,1,1,4.0,1225734739
1,1,110,4.0,1225865086
2,1,158,4.0,1225733503
3,1,260,4.5,1225735204
4,1,356,5.0,1225735119


,userId,movieId,tag,timestamp
0,10,260,good vs evil,1430666558
1,10,260,Harrison Ford,1430666505
2,10,260,sci-fi,1430666538
3,14,1221,Al Pacino,1311600756
4,14,1221,mafia,1311600746


## Ratings

### How many ratings per userId are there? 
- For collaborative (clustering) - Are there features with few observations? (Later important to filter out users and movies with few observations)

In [4]:
ratings_per_user = ratings.groupby("userId").size().reset_index(name="n_ratings")
ratings_per_movie = ratings.groupby("movieId").size().reset_index(name="n_ratings")

display(ratings_per_user["n_ratings"].describe())
display(ratings_per_movie["n_ratings"].describe())


count    330975.00000
mean        102.21969
std         232.15453
min           1.00000
25%          15.00000
50%          31.00000
75%          98.00000
max       33332.00000
Name: n_ratings, dtype: float64

count     83239.000000
mean        406.446041
std        2806.975876
min           1.000000
25%           2.000000
50%           5.000000
75%          26.000000
max      122296.000000
Name: n_ratings, dtype: float64

**Ratings per user**

This means:

- there are 330,975 users
- the median user has given only 31 ratings
- 25% of users have given at most 15 ratings
- 75% of users have given at most 98 ratings
- the mean is 102, which is much higher than the median of 31
- the maximum value, 33,332, is extremely high

*Conclusion*

The distribution is strongly right-skewed:

- most users are only mildly active
- a smaller number of users are very active
- the very active users pull the mean upward substantially

This means that if you build a collaborative filtering system, you should probably filter out users with very few ratings, since their profiles will be weak.

A reasonable first threshold could be, for example:

- at least 20 ratings
- or at least 30 ratings


**Ratings per movie**

This means:

- there are 83,239 movies with ratings
- the median movie has only 5 ratings
- 25% of movies have at most 2 ratings
- 75% of movies have at most 26 ratings
- the mean is 406, which is far higher than the median of 5
- the maximum value, 122,296, shows that a few movies are extremely popular

*Conclusion*

Here too, the distribution is extremely right-skewed:

- most movies have very few ratings
- a small number of movies have a very large number of ratings
- popular blockbuster movies dominate the overall totals

This means that you will likely want to filter out movies with very few ratings if they are to be used in recommendations or in user profiles.

A reasonable first threshold could be:

- at least 5 ratings
- at least 10 ratings
- or more, depending on how strict you want to be


## Tags 